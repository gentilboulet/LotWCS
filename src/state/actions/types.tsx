import { IInitialStateAction } from './initial';
import { IHeaderAction } from './header';
import { IChiAction } from './chi';
import { ISkillAction } from './skills';
import { ILoresheetAction } from './loresheets';
import { IHistoryAction } from './history';
import { IKungFuAction } from './kungfus';
import { IVirtueAction } from './virtues';

export type IAction =
  IInitialStateAction | IHeaderAction | ISkillAction | IHistoryAction |
  ILoresheetAction | IChiAction | IKungFuAction | IVirtueAction;
