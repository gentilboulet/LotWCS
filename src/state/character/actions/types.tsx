import { ActionType } from "typesafe-actions";

import * as chi from "./chi";
import * as header from "./header";
import * as kungfu from "./kungfu";
import * as loresheets from "./loresheets";
import * as skills from "./skills";
import * as virtues from "./virtues";

export type ICharacterAction =
  | ActionType<typeof chi>
  | ActionType<typeof header>
  | ActionType<typeof kungfu>
  | ActionType<typeof loresheets>
  | ActionType<typeof skills>
  | ActionType<typeof virtues>;
