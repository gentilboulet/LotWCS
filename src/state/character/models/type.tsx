import { TArchetype } from "../../../data/archetypes";
import { TRank } from "../../../data/ranks";
import { TAutomaticsState } from "./automatics";
import { TChiState } from "./chi";
import { TDiscountsState } from "./discounts";
import { IKungFuState } from "./kungfu";
import { ILoresheetsState } from "./loresheets";
import { TSkillsState } from "./skills";
import { TVirtuesState } from "./virtues";

export interface ICharacterState {
  name: string | undefined;
  concept: string | undefined;
  archetype: TArchetype | undefined;
  rank: TRank | undefined;

  entanglement: number;
  destiny: number;

  chi: TChiState;
  kungfu: IKungFuState;
  loresheets: ILoresheetsState;
  skills: TSkillsState;
  virtues: TVirtuesState;

  automatics: TAutomaticsState;
  discounts: TDiscountsState;
}
