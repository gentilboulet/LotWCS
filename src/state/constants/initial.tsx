export const INITIAL_STATE = 'INITIAL_STATE';
export type INITIAL_STATE = typeof INITIAL_STATE;
