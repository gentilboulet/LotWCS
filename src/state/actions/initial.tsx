import * as constants from 'constants/initial';

export interface IInitialState {
  type: constants.INITIAL_STATE;
}

export type IInitialStateAction = IInitialState;

export function resetToInitialState(): IInitialState {
  return {type: constants.INITIAL_STATE};
}
