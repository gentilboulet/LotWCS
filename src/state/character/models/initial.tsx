/* eslint-disable import/first */
import { createState as createAutomaticsState } from "./automatics";
import { createState as createChiState } from "./chi";
import { createState as createDiscountsState } from "./discounts";
import { createState as createaKungFuState } from "./kungfu";
import { createState as createLoresheetsState } from "./loresheets";
import { createState as createSkillsState } from "./skills";
import { ICharacterState } from "./type";
import { createState as createVirtuesState } from "./virtues";

export function emptyStateFactory(): ICharacterState {
  return {
    archetype: undefined,
    concept: undefined,
    name: undefined,
    rank: undefined,

    destiny: 0,
    entanglement: 0,

    chi: createChiState(),
    kungfu: createaKungFuState(),
    loresheets: createLoresheetsState(),
    skills: createSkillsState(),
    virtues: createVirtuesState(),

    automatics: createAutomaticsState(),
    discounts: createDiscountsState()
  };
}

import { setArchetype, setRank } from "../actions/header";
import { skillSpecialityBuy } from "../actions/skills";
import { playActions } from "../reducers/global";
import { zeroCost } from "./costs";

export function testingStateFactory(): ICharacterState {
  const initial = emptyStateFactory();
  const actions = [
    setRank(2),
    setArchetype("warrior"),
    skillSpecialityBuy("Awareness", "Hear", zeroCost),
    skillSpecialityBuy("Awareness", "Sight", zeroCost)
  ];
  return playActions(initial, actions);
}

export function initialStateFactory(): ICharacterState {
  return emptyStateFactory();
}
