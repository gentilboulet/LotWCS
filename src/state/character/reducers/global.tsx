import { produce } from "immer";
import { isActionOf } from "typesafe-actions";

import { ICharacterAction } from "../actions/types";
import { ICharacterState } from "../models/type";

import { initialStateFactory } from "../models/initial";

// Sub Reducers
import { chiReducer } from "./chi";
import { headerReducer } from "./header";
import { kungfuReducer } from "./kungfu";
import { loresheetsReducer } from "./loresheets";
import { skillsReducer } from "./skills";
import { virtuesReducer } from "./virtues";

import * as chi from "../actions/chi";
import * as header from "../actions/header";
import * as kungfu from "../actions/kungfu";
import * as loresheets from "../actions/loresheets";
import * as skills from "../actions/skills";
import * as virtues from "../actions/virtues";

export function isCharacterAction(action: any): boolean {
  if (isActionOf(Object.values(chi), action)) {
    return true;
  }
  if (isActionOf(Object.values(header), action)) {
    return true;
  }
  if (isActionOf(Object.values(kungfu), action)) {
    return true;
  }
  if (isActionOf(Object.values(loresheets), action)) {
    return true;
  }
  if (isActionOf(Object.values(skills), action)) {
    return true;
  }
  if (isActionOf(Object.values(virtues), action)) {
    return true;
  }
  return false;
}

export function globalReducer(
  state: ICharacterState | undefined,
  action: ICharacterAction,
): ICharacterState {
  if (!state) {
    return globalReducer(initialStateFactory(), action);
  }

  return produce(state, draftState => {
    if (isActionOf(Object.values(chi), action)) {
      draftState = chiReducer(state, action);
    }
    if (isActionOf(Object.values(header), action)) {
      draftState = headerReducer(state, action);
    }
    if (isActionOf(Object.values(kungfu), action)) {
      draftState = kungfuReducer(state, action);
    }
    if (isActionOf(Object.values(loresheets), action)) {
      draftState = loresheetsReducer(state, action);
    }
    if (isActionOf(Object.values(skills), action)) {
      draftState = skillsReducer(state, action);
    }
    if (isActionOf(Object.values(virtues), action)) {
      draftState = virtuesReducer(state, action);
    }
    return draftState;
  });
}

export function playActions(
  state: ICharacterState,
  toPlay: ICharacterAction[],
): ICharacterState {
  return produce(state, draft => {
    let newState = draft;
    toPlay.forEach(action => {
      newState = globalReducer(newState, action);
    });
    return newState;
  });
}
