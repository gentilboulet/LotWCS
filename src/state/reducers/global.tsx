import { IAction } from 'state/actions/types';
import { IStoreState } from 'state/types';

// Sub Reducers
import * as header from 'constants/header';
import { headerReducer } from 'state/reducers/header';

import * as chi from 'constants/chi';
import { chiReducer } from 'state/reducers/chi';

import * as initial from 'constants/initial';
import { initialStateFactory } from 'state/initial';

import * as kungfus from 'constants/kungfus';
import { kungfuReducer } from 'state/reducers/kungfus';

import * as history from 'constants/history';
import { historyReducer } from 'state/reducers/history';

import * as skills from 'constants/skills';
import { skillsReducer } from 'state/reducers/skills';

import * as loresheets from 'constants/loresheets';
import { loresheetsReducer } from 'state/reducers/loresheets';

import * as virtues from 'constants/virtues';
import { virtuesReducer } from 'state/reducers/virtues';

export function globalReducer(state: IStoreState, action: IAction): IStoreState {
  switch (action.type) {
    case header.HEADER_SET_ARCHETYPE:
    case header.HEADER_SET_CONCEPT:
    case header.HEADER_SET_NAME:
    case header.HEADER_SET_RANK:
      return headerReducer(state, action);
    case chi.CHI_BUY:
      return chiReducer(state, action);
    case initial.INITIAL_STATE:
      return initialStateFactory();
    case kungfus.KUNGFU_OPEN_STYLE:
    case kungfus.KUNGFU_CUSTOM_NAME_FOR_STYLE:
    case kungfus.KUNGFU_CUSTOM_NAME_FOR_TECHNIQUE:
    case kungfus.KUNGFU_BUY_TECHNIQUE:
      return kungfuReducer(state, action);
    case skills.SKILLS_BUY:
      return skillsReducer(state, action);
    case loresheets.LORESHEET_OPEN:
    case loresheets.LORESHEET_BUY_OPTION:
      return loresheetsReducer(state, action);
    case history.HISTORY_DELETE:
      return historyReducer(state, action);
    case virtues.VIRTUE_INCREASE:
      return virtuesReducer(state, action);
    default:
      return state;
  }
}
