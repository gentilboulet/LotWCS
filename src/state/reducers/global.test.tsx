import { IAction } from 'state/actions/types';
import { initialStateFactory } from 'state/initial';
import { IStoreState } from 'state/type';

import { globalReducer } from './global';

const initialState: IStoreState  = initialStateFactory();

describe('Testing globalReducer', () => {
  it('should do nothing with a junk action', () => {
    expect( initialState ).toMatchSnapshot();
    const junk = { type: 'JUNK_ACTION' };
    expect( globalReducer(initialState, junk as IAction )).toMatchObject( initialState );
  });
});
