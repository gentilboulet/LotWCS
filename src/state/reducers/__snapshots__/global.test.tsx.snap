// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing globalReducer should do nothing with a junk action 1`] = `
Immutable.Record {
  "name": "No Name",
  "concept": "No Concept",
  "archetype": "",
  "archetypeModified": false,
  "rank": "",
  "rankValue": -1,
  "rankModified": false,
  "entanglement": 0,
  "destiny": 0,
  "skills": Immutable.List [
    Immutable.Record {
      "name": "Awareness",
      "value": 0,
    },
    Immutable.Record {
      "name": "Confidence",
      "value": 0,
    },
    Immutable.Record {
      "name": "Crafting",
      "value": 0,
    },
    Immutable.Record {
      "name": "Finesse",
      "value": 0,
    },
    Immutable.Record {
      "name": "Hardiness",
      "value": 0,
    },
    Immutable.Record {
      "name": "Inspire",
      "value": 0,
    },
    Immutable.Record {
      "name": "Learning",
      "value": 0,
    },
    Immutable.Record {
      "name": "Medicine",
      "value": 0,
    },
    Immutable.Record {
      "name": "Might",
      "value": 0,
    },
    Immutable.Record {
      "name": "Perform",
      "value": 0,
    },
    Immutable.Record {
      "name": "Politics",
      "value": 0,
    },
    Immutable.Record {
      "name": "Ride",
      "value": 0,
    },
    Immutable.Record {
      "name": "Stealth",
      "value": 0,
    },
    Immutable.Record {
      "name": "Survival",
      "value": 0,
    },
    Immutable.Record {
      "name": "Tactics",
      "value": 0,
    },
    Immutable.Record {
      "name": "Wu Wei",
      "value": 0,
    },
  ],
  "skillSpecialities": Immutable.List [],
  "loresheets": Immutable.List [],
  "loresheetOptions": Immutable.List [],
  "discounts": Immutable.List [],
  "bonuses": Immutable.List [],
  "history": Immutable.List [
    Object {
      "type": "INITIAL_STATE",
    },
  ],
  "chi": Immutable.Record {
    "general": 0,
    "generalCultivation": 0,
    "wood": 0,
    "woodCultivation": 0,
    "fire": 0,
    "fireCultivation": 0,
    "earth": 0,
    "earthCultivation": 0,
    "water": 0,
    "waterCultivation": 0,
    "metal": 0,
    "metalCultivation": 0,
    "enlightened": 0,
    "enlightenedCultivation": 0,
    "demon": 0,
    "demonCultivation": 0,
  },
  "externalKungFus": Immutable.List [],
  "internalKungFus": Immutable.List [],
  "virtues": Immutable.List [
    Immutable.Record {
      "name": "Honor",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Immutable.Record {
      "name": "Benevolence",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Immutable.Record {
      "name": "Righteousness",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Immutable.Record {
      "name": "Loyalty",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Immutable.Record {
      "name": "Force",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Immutable.Record {
      "name": "Revenge",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Immutable.Record {
      "name": "Individualism",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Immutable.Record {
      "name": "Obsession",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Immutable.Record {
      "name": "Ruthlessness",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Immutable.Record {
      "name": "Ferocity",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
  ],
}
`;
